src/portable/raspberrypi/rp2040/rp2040_usb.c

```diff
@@ -294,7 +294,9 @@ bool hw_endpoint_xfer_continue(struct hw_endpoint *ep)
   // Part way through a transfer
   if (!ep->active)
   {
-    panic("Can't continue xfer on inactive ep %d %s", tu_edpt_number(ep->ep_addr), ep_dir_string);
+    pico_info("Ignore xfer on inactive ep %d %s", tu_edpt_number(ep->ep_addr), ep_dir_string[tu_edpt_dir(ep->ep_addr)]);
+    _hw_endpoint_lock_update(ep, -1);
+    return false;
```

Comment:
Returning true here would call `driver->xfer_cb()` with the closed EP, not sure if driver supports that.

My understanding is endpoints are closed in response to a control transfer on EP0, by which time all transfers on the endpoint being closed should have completed. So handling `USB_INTS_BUFF_STATUS` before `USB_INTS_SETUP_REQ` in the IRQ should prevent the transfer-after-close situation.

Reply:
If data is **actually transferred via bus**, it should be submitted to higher layer usbd/class driver. It would be their choice to whether process or ignore data. dcd shouldn't assume that.

> My understanding is endpoints are closed in response to a control transfer on EP0, by which time all transfers on the endpoint being closed should have completed. So handling USB_INTS_BUFF_STATUS before USB_INTS_SETUP_REQ in the IRQ should prevent the transfer-after-close situation.

Yes, but since dcd_edpt_close() is called within tud_task() and controller may be already in the middle of transferring by the time we process the SET_INTERFACE. It is a race, there is always chance. To be honest, I don't actually see any issue with that. 

Btw did you experience this often enough, and which callback it triggers ?

Reply:
It's a difficult topic. I've never seen a transfer event being delivered after the EP was closed. This change was a precaution based on the comment from `usb.c`. Do you agree to roll back to original `panic()`? If it comes up in practice, then it will be loud and clear and easier to deal with.

I propose also updating IRQ to handle buff_status before setup_req. This should take care of transfer completing immediately before EP close, even if there's still a chance for races like you said.

Reply:
> It's a difficult topic. I've never seen a transfer event being delivered after the EP was closed. This change was a precaution based on the comment from usb.c. Do you agree to roll back to original panic()? If it comes up in practice, then it will be loud and clear and easier to deal with.

yeah sure, if you haven't run into this scenario. It is best to leave it as it is. I have an tendency to not write code for event that isn't happened yet (to avoid having code that isn't run).

> I propose also updating IRQ to handle buff_status before setup_req. This should take care of transfer completing immediately before EP close, even if there's still a chance for races like you said.

If you could make an PR, I will be happy to review.

Reply:
Patch submitted: ae970ba2e2cc83881c52e66cd79fad5947e5e59e